# 4.5 Data Consistency Checks

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path= r'/Users/eleni_icon/09-2024 Instacart Basket Analysis'

In [3]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [4]:
df_prods=pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [6]:
df_ords.drop(columns=['Unnamed: 0'], inplace =True)

In [7]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# Mixed-type data

In [8]:
#Create a dataframe
df_test=pd.DataFrame()

In [9]:
#Create a mixed type column
df_test['mix']=['a','b',1,True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
# Check for any mixed-type columns
for col in df_test.columns.tolist():
    weird = (df_test[col].map(type) != df_test[col].iloc[0].__class__).any()
    if weird:
        print(col)

mix


In [12]:
# Convert column's data from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

# Missing values

In [13]:
# Find missing values (takes sum of Trues in a  column)
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
# Create subset of the dataframe that contains the nulls
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Addressing Missing Values

There a few ways to deal with missing data:

1. Create a new variable that acts like a flag based on the missing value.
2. Impute the value with the mean or median of the column (if the variable is numeric).
3. Remove or filter out the missing data.

In [16]:
df_prods.shape

(49693, 5)

In [17]:
# Create a new dataframe, setting the 'isnull()' condition to False so that only non-missing values are present
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
# Check the number of rows again to make sure it descreased
df_prods_clean.shape

(49677, 5)

In [19]:
# An alternative method to drop all missing values is via the following command
df_prods.dropna(inplace = True)

In [20]:
# To drop only NaNs from a particular column use the following command
df_prods.dropna(subset = ['product_name'], inplace = True)

In [21]:
#To drop all missing values --> df_prods.dropna(inplace = True)

#To drop only the NaNs from a particular column --> df_prods.dropna(subset = [‘product_name’], inplace = True)

# Duplicates

In [22]:
# Create subset of df_prods_clean that contains only rows of duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
# Check current number of rowns in df_prods_clean
df_prods_clean.shape

(49677, 5)

In [25]:
# Create datafram that doesn't include the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
df_prods_clean_no_dups.shape

(49672, 5)

# Exporting changes

In [27]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5

# Step 2

In [28]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

The missing values in days_since_prior_order, the minimum value of 0 in days_since_prior_order, and the high order numbers are areas that should be investigated further.

# Step 3

In [29]:
for col in df_ords.columns.tolist():
    weird = (df_ords[col].map(type) != df_ords[col].iloc[0].__class__).any()
    if weird:
        print(col)


order_id
user_id
order_number
orders_day_of_week
order_hour_of_day
days_since_prior_order


In [30]:
# There's no mixed data present in the df_ords dataframe
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [31]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(5)
memory usage: 156.6 MB


# Step 5

In [32]:
# Find missing values (takes sum of Trues in a  column)
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

As expected, is the 'days_since_prior_order' has 206209 values missing.

# Step 6

In [33]:
# Create subset of the dataframe that contains the nulls
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [34]:
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [35]:
# Filter the dataframe where 'order_number' equals 1
df_first_orders = df_ords_nan[df_ords_nan['order_number'] == '1']

In [36]:
# Check if all of these have missing values in 'days_since_prior_order'
df_first_orders_missing = df_first_orders['days_since_prior_order'].isna().all()

In [37]:
# Print the result
print("All first orders have missing 'days_since_prior_order':", df_first_orders_missing) 

All first orders have missing 'days_since_prior_order': True


In [38]:
# Address the missing values by creating a column that flags order numbers as either True/False depending on if the 'order_number' = 1
df_ords['first_order'] = df_ords['days_since_prior_order'].isna()

In [39]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
1                        7                    15.0        False  
2                       12                    21.0        False  
3                        7                    29.0        False  
4                       15                    28.0        False  
...                    ...                     ...          ...  
3421078                 18                    29.0        False  
3421079                 10                    30.0        False  
3421080                 12                    18.0        False  
3421081                 12                     7.0        False  
3421082                 14                    30.0        False  

[3421083 rows x 7 columns]

# Step 7

In [40]:
# Create subset of df_ords that contains only rows of duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [41]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

 No duplicates found;if there were duplicates I would execute '.drop_duplicates()' to creat a new datafreame without duplicates.

In [42]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [43]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))

In [44]:
# Check number of unique product ids
df_unique_product_ids = df_prods_clean_no_dups['product_id'].nunique()

In [45]:
print(f"Number of unique product IDs: {df_unique_product_ids}")

Number of unique product IDs: 49670


In [46]:
# Check if product_id values are sequential and without gaps
df_all_ids = set(range(1, df_prods_clean_no_dups['product_id'].max() + 1))

In [47]:
# Create a set of actual ids in the dataframe
df_actual_ids = set(df_prods_clean_no_dups['product_id'].unique())

In [48]:
# Find any missing ids
missing_ids = df_all_ids - df_actual_ids

In [49]:
if missing_ids:
    print(f"There are missing product IDs: {sorted(missing_ids)}")
else:
    print("No missing product IDs; they are sequential.")

There are missing product IDs: [34, 69, 116, 262, 525, 1511, 1780, 2240, 2586, 3159, 3230, 3736, 4283, 4790, 6799, 26519, 38183, 40440]


In [50]:
outliers = df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] == 99999]

In [51]:
print(outliers)

       product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0


In [52]:
df_prods_extra_clean_no_dups = df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] != 99999]

In [53]:
df_prods_extra_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49671.000000  49671.000000   49671.000000  49671.000000
mean   24850.172334     67.762115      11.728856      7.980256
std    14340.795118     38.316100       5.850806     66.952504
min        1.000000      1.000000       1.000000      1.000000
25%    12432.500000     35.000000       7.000000      4.100000
50%    24850.000000     69.000000      13.000000      7.100000
75%    37268.500000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  14900.000000

In [54]:
df_prods_sorted = df_prods_extra_clean_no_dups.sort_values(by='prices', ascending=False)

In [55]:
top_priced_products = df_prods_sorted.head()

In [56]:
print(top_priced_products)

       product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
9020         9020  Boneless Skinless Chicken Thighs        35             12   
40490       40486                   Chicken Tenders        49             12   
21468       21467            Wild Caught Raw Shrimp        15             12   
25580       25579     Naturally Smoked Trout Fillet        15             12   

        prices  
21554  14900.0  
9020      25.0  
40490     25.0  
21468     25.0  
25580     25.0  


In [57]:
df_prods_extra_clean_no_dups_2 = df_prods_extra_clean_no_dups[df_prods_extra_clean_no_dups['prices'] != 14900]

In [58]:
df_prods_extra_clean_no_dups_2.describe()

product_id      aisle_id  department_id        prices
count  49670.000000  49670.000000   49670.000000  49670.000000
mean   24850.238716     67.761305      11.728770      7.680437
std    14340.931850     38.316060       5.850834      4.199381
min        1.000000      1.000000       1.000000      1.000000
25%    12432.250000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.750000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

# Step 9

In [59]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.csv'))

In [60]:
df_prods_extra_clean_no_dups_2.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))